[FSDP][optim_state_dict] Call synchronize() to ensure DTensors.to_local() is synchronized

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117799

If a tensor is converted from DTensor (to_local()), there may be some async communication that has not finished yet. Calling `clone()` with that tensor does not seem to work (and may increase the memory usage, users report OOM). This is a temporary fix.

Differential Revision: [D52890462](https://our.internmc.facebook.com/intern/diff/D52890462/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

fegin committed:
[FSDP][optim_state_dict] Call synchronize() to ensure the temporary tensors being recycled

emporary tensors could not be recycled unless the operations are finished.  Calling synchronize() can ensure all the operations are finished. The action can prevent OOM from happening.

Differential Revision: [D52890462](https://our.internmc.facebook.com/intern/diff/D52890462/)

[ghstack-poisoned]

awgu commented:
Does this part of the optim state dict load use multiple streams (which is why memory is not freed immediately when there are no more references)?

fegin commented:
No multiple streams. But even if the temporary tensors are not referenced in Python, it can still being used, iiuc. For example, the `clone()` one will keep the source tensor (and its memory) alive until the operation is done. Or do you think this should not happen?

awgu commented:
I was wondering if you have tried `del`-ing the source that is being cloned and seeing if that frees it at the time of `del`.

Suppose at first we have this:
```
a = torch.empty((3,), device="cuda")
b = a.clone()
c = torch.empty((3,), device="cuda")
```
`c` cannot reuse the memory of `a` since `a` is still alive (due to the Python reference).

Now, suppose we `del a`:
```
a = torch.empty((3,), device="cuda")
b = a.clone()
del a  # <--- add this
c = torch.empty((3,), device="cuda")
```
`c` can reuse the memory of `a`. It does not need to wait until the GPU copy kernel from `clone()` finishes because any subsequent GPU kernel using the memory for `c` will be sequentially after the GPU copy kernel.

I did not see any special handling of `clone` that would record a `cudaEvent` and free the memory later (like in the case of multiple streams and `recordStream`). I would be curious if `del`-ing did not free the memory instantly if there is only a single stream.

<details>
<summary> clone implementation </summary>

https://github.com/pytorch/pytorch/blob/2f84a9d37ced9ccc0b5ff57c93f31f58ad44604d/aten/src/ATen/native/TensorFactories.cpp#L1730-L1751
</details>

fegin commented:
@awgu that makes sense. However, the tensor should already be deleted as it has no reference after the inner util function. So I don't think `del` makes any difference of that. I would do more tests to verify.

fegin commented:
@awgu I can confirm that `del` alone is not enough. I'll check if multiple streams are being used.

fegin committed:
Update on "[FSDP][optim_state_dict] Call synchronize() to ensure the temporary tensors being recycled"


emporary tensors could not be recycled unless the operations are finished.  Calling synchronize() can ensure all the operations are finished. The action can prevent OOM from happening.

Differential Revision: [D52890462](https://our.internmc.facebook.com/intern/diff/D52890462/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fegin commented:
@awgu The root cause is  `DTensor.to_local()` has asynchronous communication by default. So I believe we need to call `synchronize()` in such a case.

awgu commented:
> @awgu The root cause is `DTensor.to_local()` has asynchronous communication by default. So I believe we need to call `synchronize()` in such a case.

Thanks for looking into this!

@wanchaol Is there any way to specify to `to_local()` to use synchronous collectives? We want to avoid `recordStream`, and IIUC, we only avoid it for synchronous collectives. (cc: @kwen2501).

If we can run synchronous collectives and avoid `recordStream`, then I think we can avoid the CPU sync from `synchronize()` and instead just rely on the PG NCCL calling `current_stream.wait_stream(nccl_stream)`. Avoiding the CPU sync here _could_ prevent some CPU boundedness (though I have not looked at any profiles).

fegin commented:
@awgu This is optimizer state_dict code path. What's the down side of using `torch.cuda.synchronize()`?

awgu commented:
@fegin The downside is just performance from synchronizing the CPU. If there is not much performance downside, then using `synchronize` sounds good to me!

(I do not have a good mental model of the optimizer state dict performance, so maybe this is not really an issue. However, in general, I do think that considering the performance of state dict makes sense.)

fegin commented:
@awgu Yup, agree that performance is important. Currently, `synchronize()` is probably not the main bottleneck. The main bottleneck is the allgather which is done in a per-parameter way without batching. However, it is hard to batching the allgather for optimizer state dict load.

fegin committed:
Update on "[FSDP][optim_state_dict] Call synchronize() to ensure the memory used by DTensors.to_local() being recycled"


If a tensor is converted from DTensor (to_local()), there may be some async communication that has not finished yet. Calling synchronize() can ensure all the operations are finished. The action can prevent OOM from happening.

Differential Revision: [D52890462](https://our.internmc.facebook.com/intern/diff/D52890462/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fegin commented:
This becomes a bigger issue than just a performance issue. If we don't call `torch.cuda.synchronize()`, the subsequent `clone()` will not work -- the value will not be correctly cloned. 

cc., @wanchaol @awgu @wz337 @LucasLLC

fegin committed:
Update on "[FSDP][optim_state_dict] Call synchronize() to ensure DTensors.to_local() is synchronized"


If a tensor is converted from DTensor (to_local()), there may be some async communication that has not finished yet. Calling `clone()` with that tensor does not seem to work (and may increase the memory usage, users report OOM). This is a temporary fix.

Differential Revision: [D52890462](https://our.internmc.facebook.com/intern/diff/D52890462/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 commented:
Can we use stream sync instead of device sync?

LucasLLC reviewed (approved):
Lacking a bit of context but approving to unblock the fix

kwen2501 commented:
if only device sync works and stream sync doesn't, that means there is something wrong with DTensor's to_local() -- it should sync its communication work back to the "current stream" (or else provide a Work handle for user to sync at some point)

fegin commented:
@kwen2501 `_optim_utils.py` calls to `DTensor.to_local()` to gather the tensor. It does not know which stream `DTensor.to_local()` uses. So I  don't think I'm able to use stream wait.

kwen2501 commented:
If `to_local()`'s API does not have flags like `async_op=True|False`, it **must** always sync back to main stream, so that you don't need to figure out which stream to wait. Said plainly, `to_local()` must always call `work.wait()` inside or expose `work` out.

awgu reviewed (approved):
Landing this as a temporary fix sounds good to me! 

We should separately figure out how to make `AsyncCollectiveTensor` wait before `clone()` (if that is root issue).

wconstab commented:
in the more general sense of a user using DTensor.to_local, i don't know if it makes sense that operation has to be synchronous.  Wouldn't it be reasonable to expect a to_local call to return a new 'ACT' that represents ongoing tensor work?

wanchaol commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -403,8 +404,17 @@ def _shard_orig_param_state(
-            value = value.flatten()[intra_param_start_idx : intra_param_end_idx + 1].clone()  # type: ignore[operator]
+            if fsdp_state._device_mesh is not None:
+                # We have to call synchronize() if the tensor is gathered from
+                # DTensor.  Otherwise, the later `clone()` will cause errors.
```
are you using the `full_tensor()` API to do the gathering or something else? iirc `full_tensor` gives sync behavior

wz337 replied:
> are you using the `full_tensor()` API to do the gathering or something else? iirc `full_tensor` gives sync behavior

No. We are still using redistribute for the all_gather, since `full_tensor()` API was introduced later. 
Maybe we could update this line to use `full_tensor` API? https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_optim_utils.py#L1455

kwen2501 commented:
> Wouldn't it be reasonable to expect a to_local call to return a new 'ACT' that represents ongoing tensor work?

The reasonableness depends on how much tendency users have to write
`x.to_local()`
vesus
`x = x.to_local()`

In a world that `tensor.to(...)` is prevailing, I'd say the tendency is non-negligible...

awgu commented:
@kwen2501 `tensor.to(...)` is not inplace though. Only `nn.Module.to` is in-place.

kwen2501 commented:
@awgu You are right. Thanks for the correction!

fegin commented:
I have not found the root cause. Since `_gather_state_dict()` is used not only by FSDP but also PP-FSDP, I decide to change the behavior of  `_gather_state_dict()` to always call `wait()` before returning the tensor. The new PR is https://github.com/pytorch/pytorch/pull/118197.
